src/systems/model_photo_shoot/ModelPhotoShoot.cc

```diff
@@ -0,0 +1,283 @@
+  public: void SavePicture (const ignition::rendering::CameraPtr _camera,
+                    const ignition::math::Pose3d _pose,
+                    const std::string _fileName) const;
+
+  /// \brief Name of the loaded model.
+  public: std::string modelName;
+
+  /// \brief model
+  public: std::shared_ptr<ignition::gazebo::Model> model;
+
+  /// \brief model world pose
+  public: ignition::math::Pose3d modelPose3D;
+
+  /// \brief Connection to pre-render event callback.
+  public: ignition::common::ConnectionPtr connection{nullptr};
+
+  /// \brief Boolean to control we only take the pictures once.
+  public: bool takePicture{true};
+
+  /// \brief Boolean to control if joints should adopt random poses.
+  public: bool randomPoses{false};
+
+  /// \brief File to save translation and scaling info.
+  public: std::ofstream savingFile;
+};
+
+//////////////////////////////////////////////////
+ModelPhotoShoot::ModelPhotoShoot()
+    : System(), dataPtr(std::make_unique<ModelPhotoShootPrivate>())
+{
+}
+
+//////////////////////////////////////////////////
+void ModelPhotoShoot::Configure(const ignition::gazebo::Entity &_entity,
+                                const std::shared_ptr<const sdf::Element> &_sdf,
+                                ignition::gazebo::EntityComponentManager &_ecm,
+                                ignition::gazebo::EventManager &_eventMgr)
+{
+  std::string save_data_location =
+      _sdf->Get<std::string>("translation_data_file");
+  if (save_data_location.empty())
+  {
+    igndbg << "No data location specified, skipping translaiton data"
+              "saving.\n";
+  }
+  else
+  {
+    igndbg << "Saving translation data to: "
+        << save_data_location << std::endl;
+    this->dataPtr->savingFile.open(save_data_location);
+  }
+
+  if (_sdf->HasElement("random_joints_pose"))
+  {
+    this->dataPtr->randomPoses = _sdf->Get<bool>("random_joints_pose");
+  }
+
+  this->dataPtr->connection =
+      _eventMgr.Connect<ignition::gazebo::events::PostRender>(std::bind(
+          &ModelPhotoShootPrivate::PerformPostRenderingOperations,
+          this->dataPtr.get()));
+
+  this->dataPtr->model = std::make_shared<ignition::gazebo::Model>(_entity);
+  this->dataPtr->modelName = this->dataPtr->model->Name(_ecm);
+  // Get the pose of the model
+  this->dataPtr->modelPose3D =
+      ignition::gazebo::worldPose(this->dataPtr->model->Entity(), _ecm);
+}
+
+//////////////////////////////////////////////////
+void ModelPhotoShoot::PreUpdate(
+    const ignition::gazebo::UpdateInfo &,
+    ignition::gazebo::EntityComponentManager &_ecm)
+{
+  if (this->dataPtr->randomPoses)
+  {
+    std::vector<gazebo::Entity> joints = this->dataPtr->model->Joints(_ecm);
+    unsigned seed =
+        std::chrono::system_clock::now().time_since_epoch().count();
+    std::default_random_engine generator(seed);
+    for (auto joint = joints.begin(); joint != joints.end(); ++joint)
```

Comment:
nit: Easier to use range-based for loops
```suggestion
    for (const auto &joint = joints)
```